core/src/test/java/org/infinispan/topology/MixedVersionClusterTest.java

```diff
@@ -0,0 +1,205 @@
+         globalBuilder.serialization().addContextInitializer(RpcSCI.INSTANCE);
+         addCacheManager(i, globalBuilder, newVersion);
+         waitForClusterToForm(TEST_CACHE);
+
+         if (i != 0) {
+            // Assert that all nodes are aware we're in a mixed cluster
+            for (int j = 0; j < NUM_NODES; j++) {
+               assertTrue(isMixedCluster(j));
+               assertEquals(NodeVersion.INSTANCE, getOldestMember(j));
+            }
+            assertCommandFailsOnAllManagers(cmd);
+         } else {
+            // Assert that all nodes are aware we're no longer in a mixed cluster
+            for (int j = 0; j < NUM_NODES; j++) {
+               assertFalse(isMixedCluster(j));
+               assertEquals(newVersion, getOldestMember(j));
+            }
+            assertCommandPassesOnAllManagers(cmd);
+         }
+      }
+   }
+
+   private void assertCommandFailsOnAllManagers(CacheRpcCommand cmd) throws Exception {
+      for (int i = 0; i < managers().length; i++) assertCommandOnManager(i, cmd, true);
+   }
+
+   private void assertCommandPassesOnAllManagers(CacheRpcCommand cmd) throws Exception {
+      for (int i = 0; i < managers().length; i++) assertCommandOnManager(i, cmd, false);
+   }
+
+   private void assertCommandOnManager(int i, CacheRpcCommand cmd, boolean fail) throws Exception {
+      var rpcManager = manager(i).getCache(TEST_CACHE).getAdvancedCache().getRpcManager();
+      var syncOpts = rpcManager.getSyncRpcOptions();
+
+      var address = rpcManager.getMembers().stream().filter(a -> !a.equals(rpcManager.getAddress())).findFirst().get();
+      assertCommandExec(
+            rpcManager.invokeCommand(address, cmd, MapResponseCollector.ignoreLeavers(), syncOpts), fail
+      );
+
+      var targets = rpcManager.getMembers();
+      assertCommandExec(
+            rpcManager.invokeCommand(targets, cmd, MapResponseCollector.ignoreLeavers(), syncOpts), fail
+      );
+
+      assertCommandExec(
+            rpcManager.invokeCommandOnAll(cmd, MapResponseCollector.ignoreLeavers(), syncOpts), fail
+      );
+
+      assertCommandExec(
+            rpcManager.invokeCommandStaggered(targets, cmd, MapResponseCollector.ignoreLeavers(), syncOpts), fail
+      );
+
+      assertCommandExec(
+            rpcManager.invokeCommands(targets, a -> cmd, MapResponseCollector.ignoreLeavers(), syncOpts), fail
+      );
+
+      assertCommandExec(() -> rpcManager.sendTo(address, cmd, DeliverOrder.NONE), fail);
+      assertCommandExec(() -> rpcManager.sendToMany(targets, cmd, DeliverOrder.NONE), fail);
+      assertCommandExec(() -> rpcManager.sendToAll(cmd, DeliverOrder.NONE), fail);
+   }
+
+   private void assertCommandExec(Runnable r, boolean expectFail) {
+      try {
+         r.run();
+         if (expectFail) fail();
+      } catch (CacheException e) {
+         if (!expectFail) fail("Unexpected exception: " + e.getCause());
+         assertUnsupportedException(e.getCause());
+      }
```

Comment:
don't need to reinvent the wheel. You could use this method when `expectFail==true`: `
org.infinispan.commons.test.Exceptions#expectException(java.lang.Class<? extends java.lang.Throwable>, java.lang.String, org.infinispan.commons.test.ExceptionRunnable)`